File: solutions/0063_Unique_Paths_II/dp_2D.java
class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int m = obstacleGrid.length;
        int n = obstacleGrid[0].length;
        int[][] dp = new int[m][n];
        // if left-top grid is 1, stock 0
        dp[0][0] = 1 - obstacleGrid[0][0];
        for (int j = 1; j < n; ++j) { // first row
            if (obstacleGrid[0][j] == 0) dp[0][j] = dp[0][j-1];
        }
        for (int i = 1; i < m; ++i) { // first column
            if (obstacleGrid[i][0] == 0) dp[i][0] = dp[i-1][0];
        }
        for (int i = 1; i < m; ++i) {
            for (int j = 1; j < n; ++j) {
                if (obstacleGrid[i][j] == 0) {
                    dp[i][j] = dp[i-1][j] + dp[i][j-1];
                }
            }
        }
        return dp[m-1][n-1];
    }
}

File: solutions/0221_Maximal_Square/dp_2d.py
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        if not matrix or not matrix[0]:
            return 0
        m, n = len(matrix), len(matrix[0])
        dp = [[0 for _ in range(n)] for _ in range(m)]
        best = 0

        # first row
        for j in range(n):
            dp[0][j] = int(matrix[0][j])
            best = max(best, dp[0][j])
        # first column
        for i in range(m):
            dp[i][0] = int(matrix[i][0])
            best = max(best, dp[i][0])
        # dp from [1,m] & [1,n]
        for i in range(1, m):
            for j in range(1, n):
                dp[i][j] = 0 if matrix[i][j] == '0' else min(dp[i-1][j-1], dp[i-1][j], dp[i][j-1]) + 1
                best = max(best, dp[i][j])
        return best*best
